Refactoring of the existing security-secrets-setup code

jim-wang-intel opened the pull request:
Refactor the existing code to accept new changes for pki-init integration and better testable code

  - Abstract os.Exit() call so it can be tested without losing code coverage
  - Abstract pki-init option so that it can accept the pki-init options later
  - Change the local logging client to use global one so that it is more consistent with security-proxy and the logger can be used through out all options in this service
  - Refactor the execution of -configFile into function so that it can be integrated later with better generation option
  - Add unit test for main.go for increasing test coverage from the origin
  - Add unit test for x509.go for increasing test coverage from the origin

Related to: Issue #1635 

Signed-off-by: Jim Wang <yutsung.jim.wang@intel.com>

jim-wang-intel commented on cmd/security-secrets-setup/main.go:
```diff
@@ -16,80 +17,98 @@ package main
+	optsExecutor, statusCode, err := setupPkiInitOption()
 	if err != nil {
-		log.Fatalf("ERROR: %s: %s", msg, err) // fatalf() =  Prinf() followed by a call to os.Exit(1)
+		return
```
@mkbhanda - that is a good catch, should return error with status code

jim-wang-intel replied:
yea, this `run()` returns error and error is logged on line 74 above

jim-wang-intel commented on .gitignore:
```diff
@@ -26,6 +26,7 @@ cmd/sys-mgmt-agent/sys-mgmt-agent
 cmd/support-scheduler/support-scheduler
 cmd/sys-mgmt-executor/sys-mgmt-executor
 docs/_build/
+cmd/security-secrets-setup/config/*
```
ok

jim-wang-intel commented:
> 
> 
> pkiOption_test.go file may need "gofmt -w -l -s" . Other than that it looks good to me.

@tingyuz Tingyu, I ran that, and gofmt tool reports clean to me on that. so we are fine.

jim-wang-intel commented:
@anonymouse64 Ian, any feedback? I am expecting to merge this onto master pretty soon.

anonymouse64 commented:
>@anonymouse64 Ian, any feedback? I am expecting to merge this onto master pretty soon.

I will try to review this today

anonymouse64 commented on cmd/security-secrets-setup/main_test.go:
```diff
@@ -0,0 +1,138 @@
+}
+
+func TestMainWithConfigFileOption(t *testing.T) {
+	tearDown := setupTest(t)
```
this kind of test usually reads much cleaner as:

```go
defer mockExitInstance(newArgs(), newTestExit(), newTestDispatcher())
```

where `mockExitInstance` looks like this:

```go
func mockExitInstance(mockedExiter exit) func() {
    // save for restoring
    prevExitInstance = exitInstance

    // mock the new one
    exitInstance = mockedExiter

    return func() {
        exitInstance = prevExitInstance

        // you can also do cleanup here too
        os.RemoveAll("./config")
    }
}
```

repeating this pattern for all the things you need to mock

jim-wang-intel replied:
Well, my original real purpose to do this is just for prepare Initialization/reset and tearDown for unit tests and not really for mocking purpose per se. But your idea to do this way is interesting for mocking things for other testing purpose as well.  The real mocking is actually achieved via combination of interface and testing struct like I do here.

anonymouse64 commented on cmd/security-secrets-setup/main.go:
```diff
@@ -16,80 +17,98 @@ package main
 
-// TODO: ELIMINATE THIS ----------------------------------------------------------
-func fatalIfErr(err error, msg string) {
+func setupPkiInitOption() (executor option.OptionsExecutor, status int, err error) {
```
this function doesn't do much and is only used inside the `run()` method so I think you should just roll it into `run()`. Where you use it in the tests can be similarly rolled inside, running a single function with a default value struct doesn't really warrant sharing code like this

jim-wang-intel replied:
Well, this way is more unit-testable. So someone can just unit-test this setup to make usre it does what it needs to do and not worry about the driver part (which is `run()`).  As a matter of fact, in the unit tests, `run()` is overridden and used for test cases.

anonymouse64 replied:
this specific function is not unit tested though, `run()` is what you are unit testing, this function doesn't do anything except call one other function

jim-wang-intel replied:
Adding unit test to it

anonymouse64 replied:
sorry the point I was making is that you don't need to unit test this function because it isn't doing much of anything, it's like defining this function:

```go
func doAThing() {
    return doAnotherthing(struct{})
}
```

there's nothing gained by this function and so it doesn't need a test and the function doesn't need to exist is all I'm saying, apologies if that was unclear

jim-wang-intel replied:
Ian, there is something being doing in line 107 `option.NewPkiInitOption(opts)`, this function is doing something.  As a matter of fact, this one serves as the purpose of regression-testing later on when `-generate`, `-cache`, and `-import` option are going to be added.  With this as separate function, then it can be regression tested by the unit test to see if anything break when options are added later.

anonymouse64 commented on cmd/security-secrets-setup/main_test.go:
```diff
@@ -0,0 +1,138 @@
+	main()
+}
+
+func runWithConfigFileOption() {
```
why not just roll this into the `TestMainWithConfigFileOption` directly? it's not used anywhere else and would make the tests you write easier to follow without having to jump around to all the helper functions

jim-wang-intel replied:
I am fine with that, but it is just code readability concern: with this func name it seems to read better than all codes inside this.

anonymouse64 commented on cmd/security-secrets-setup/main_test.go:
```diff
@@ -0,0 +1,138 @@
+	main()
+}
+
+func printCommandLineStrings(strs []string) {
```
it seems like this entire function could just be `fmt.Println("command line strings:",strings.Join(strs," "))`

jim-wang-intel replied:
ok

anonymouse64 commented on cmd/security-secrets-setup/main.go:
```diff
@@ -16,80 +17,98 @@ package main
-func main() {
-	start := time.Now()
-	var configFile string
+type exit interface {
```
I dislike the name of this interface, it's not clear what it does and isn't idiomatic go naming, I would like this better:

```go
type exiter interface {
    Exit()
}
```

jim-wang-intel replied:
It does sound better to me, will change it.

anonymouse64 commented on cmd/security-secrets-setup/main.go:
```diff
@@ -16,80 +17,98 @@ package main
+	callExit(int)
+}
+
+type exitCode struct{}
```
instead of defining a whole new type and interface for testing this kind of thing, what I've seen done in snapd is:

```go
var (
    ExitImpl = os.Exit
)

func main() {
    exitImpl(0)
}
```

then in tests you do this:

```go
func makeTestExit(exitedStatus *int) func(status int) {
    return func(status int){ *exitedStatus = status }
}

func TestTheThings() {
    exitCode := 0
    defer mockExit(makeTestExit(&exitCode))
    runMainWithErrorThings()
    if exitCode == 1 {
        // test pass
    } else {
        // test fail
    }
}

func mockExit(mockedExit func(status int)) func() {
    oldExit := pkg.ExitImpl
    pkg.ExitImpl = mockedExit
    return func() {
        pkg.ExitImpl = oldExit
    }
}
```

bnevis-i replied:
First class functions!

jim-wang-intel replied:
@bnevis-i  can you elaborate it more? What do you mean by "first class" functions?

jim-wang-intel replied:
After some thinking, IMHO, I like interface approach better: the reasons are
1. Interface is clear to convey the idea what type of actions you are trying to do
2. Interface is also easier to put all collections of actions to do in one contract for the caller; though in this case we only have one but with that, it is easier to override the behavior in test and in other places if you find to do so.
3. IMHO, interface is more object-oriented than just using the variables

Moreover, like I said previously, the main purpose of setup and teardown is not to do the mocking but rather to initialize and clean up the things tests needed.

anonymouse64 commented on cmd/security-secrets-setup/main_test.go:
```diff
@@ -0,0 +1,138 @@
+	defer tearDown(t, origArgs)
+	assert := assert.New(t)
+
+	runWithNoOption()
```
ditto, I think you should just roll this function into here

jim-wang-intel replied:
same concern- code readability

anonymouse64 commented on internal/pkg/config/x509_test.go:
```diff
@@ -15,30 +16,52 @@
 }
+
+func TestPkiCAOutDir(t *testing.T) {
+	x509Config, readErr := NewX509Config(testJSONFileName)
```
I prefer to always just have `err` as the error variable unless you have a specific need to juggle around multiple errors simultaneously and independently

```suggestion
	x509Config, err := NewX509Config(testJSONFileName)
```

jim-wang-intel replied:
fine with me

anonymouse64 commented on internal/security/setup/option/exitCode.go:
```diff
@@ -0,0 +1,33 @@
+package option
+
+// ExitCode is the code used for exit status
+type exitCode int
```
why have a new type for the exit code and not just define constants for the various exit codes we use for specific situations? seems like an unnecessarily abstraction

jim-wang-intel replied:
It is less verbose in the constant name without losing the context- if only defined in constant, then one would end up with 
```
const(
   normalExitCode = 0
   noOptionSelectedExitCode = 1
   exitWithErrorExitCode = 2
)
```

anonymouse64 commented on internal/pkg/config/x509_test.go:
```diff
@@ -15,30 +16,52 @@
-				return
+			x509Config, err := NewX509Config(tt.file)
+
+			if !tt.expectError {
```
this would read better as 

```go
			if tt.expectError {
				assert.NotNil(err, "expected error, none returned")
				assert.Empty(x509Config)
                        } else {
				assert.Nil(err, "unexpected error %v", err)
				assert.NotEmpty(x509Config)
                        }
```

jim-wang-intel replied:
good point

anonymouse64 commented on internal/pkg/config/x509_test.go:
```diff
@@ -15,30 +16,52 @@
+	"github.com/stretchr/testify/assert"
 )
 
+var testJSONFileName = "./res/pkisetup-test.json"
```
can we move this to testdata instead of res?

jim-wang-intel replied:
ok

anonymouse64 commented on internal/security/setup/option/pkiOption.go:
```diff
@@ -0,0 +1,54 @@
+	return statusCode.intValue(), err
+}
+
+func (pkiInitOpt *PkiInitOption) executeOptions(executors ...func(*PkiInitOption) (exitCode, error)) (exitCode, error) {
```
what's the use case for having multiple executors?

jim-wang-intel replied:
for testing purpose

jim-wang-intel commented on cmd/security-secrets-setup/main.go:
```diff
@@ -16,80 +17,98 @@ package main
+	if len(os.Args) < 2 {
+		fmt.Println("Please specify option for " + option.SecuritySecretsSetup)
+		flag.PrintDefaults()
+		exitInstance.callExit(1)
```
@anonymouse64 Ian, I am not against the exit code 0 for --help or -h option.  So it is a matter how we treat no option is provided at all. Currently we treat no option provided as exit code 1 instead of 0 but I am open to that.  @bnevis-i  what do you think?

bnevis-i replied:
I have seen Unix utilities that do it both ways.  Personally I think if the command didn't do its intended job it should return a non-zero exit status, so that thinks like exec.Command will flag the error.  But since there is no generally accepted standard, if Ian says zero status when displaying usage is the EdgeX convention, I can go along with that.

jim-wang-intel replied:
ok, i am ok to go with 0 status code then

jim-wang-intel commented on internal/security/setup/option/constant.go:
```diff
@@ -0,0 +1,22 @@
+
+const (
+	// SecuritySecretsSetup is the name of this module
+	SecuritySecretsSetup = "security-secrets-setup"
```
just to reduce the hard-coded same string in multiple places even only in main.

jim-wang-intel commented on cmd/security-secrets-setup/main.go:
```diff
@@ -16,80 +17,98 @@ package main
-	if err != nil {
-		lc.Error(err.Error())
+	if err := setup.Init(); err != nil {
+		exitInstance.callExit(1)
```
Sounds ok for me to put some comments in code for that- since the original code already does that logging inside the call, so I'd rather leave it there for less changes.

jim-wang-intel commented:
Thanks for your valuable comments and feedback, @anonymouse64 Ian, I put some replies to your comments and have addressed your PR feedback.  Please check it. Thanks! -Jim

jim-wang-intel commented on internal/security/setup/init.go:
```diff
@@ -15,27 +15,44 @@
 
-func Init() {
+func Init() error {
 	// Unfortunately I have to do this because of utilization of the LoggingClient in config.LoadFromFile below.
```
That is original old comment there.. I agree with you that Trevor should do an issue for that.

jim-wang-intel commented on internal/security/setup/option/optionHelpers.go:
```diff
@@ -0,0 +1,65 @@
+// GenTLSAssets generates the TLS assets based on the JSON configuration file
+func GenTLSAssets(jsonConfig string) error {
+	// Read the Json config file and unmarshall content into struct type X509Config
+	x509Config, readErr := config.NewX509Config(jsonConfig)
```
good catch i'll change it

jim-wang-intel commented:
@anonymouse64  Ian, I have rebased and squashed to the single commit. Thanks for great feedbacks. This one commit should be ready to merge onto master.

jim-wang-intel commented:
Hi @mkbhanda Malini, if changes are good for you, please approve it. Thanks and have a nice weekend! 
-Jim W.

mkbhanda commented on .gitignore:
```diff
@@ -19,11 +19,12 @@ cmd/core-metadata/core-metadata
 cmd/export-client/export-client
 cmd/export-distro/export-distro
 cmd/security-secrets-setup/security-secrets-setup
+cmd/security-secrets-setup/config/*
```
nit: a little out of order
